### ISSUE TITLE
at-exp reader indentation for DrRacket is buggy

### ISSUE THREAD

**Issue by LeifAndersen** [racket/racket#1240]
The indentation for at-expressions in DrRacket appears to be buggy. Namely, there seems to be three problems.

1. When I put text on a new, empty line, it will not indent until there is actually text on the line.

So if I have the following file:

```
#lang scribble/lncs

@emph{
```

And I hit enter at the end, it will put the cursor right below the opening `{`.

However, if I write out text, and hit tab again, it will move the text to where I would expect, namely:

```#lang scribble/lncs

@emph{
  some text}
```

2. When I have text on the first line of an at-expression, it doesn't move the starting text to the open curly brace.

So for example, if I have autotab the following file, it looks like this:

```
#lang scribble/lncs

@emph{Some more
 text}
```

I would expect the text to look like this:

```
#lang scribble/lncs

@emph{Some more
      text}
```

3. If the program has any non-command curly braces (that appear in the text), it uses those curly braces for indenting, rather than the current command curly brace it is in. This stays the case even after the curly brace has been closed, or if another at-exp has finished.  For example, the following is how a file is typed (before I highlight and auto-indent the whole already typed file):

```
#lang scribble/lncs

@emph{An internal {wrong}
                  indent.}
```

**Comment by rfindler** [comment]
Thank you for finding these helpful test cases!

The code is in this other repo, however: https://github.com/racket/gui/blob/master/gui-lib/scribble/private/indentation.rkt

**Comment by rfindler** [comment]
Oh, but I don't agree about this example:

```
#lang scribble/lncs

@emph{Some more
 text}
```

I think this is the correct indentation. Consider this example:

```
@paragraph{This is me typing some stuff and then I want two @bold{bold
 words}. I don't want the "word" to appear way left.}
```

**Comment by LeifAndersen** [comment]
Oh derp...facepalm. Sorry, for some reason I thought I was in that repo. Oh well, thanks anyway.

As for your third example, I guess that's fair. It also looks like the semantics of at-expressions are also what I would hope for. (Namely, it doesn't put a space before the ` word` on the second line.

Although I should point out that the default auto-indent for the expression you posted there is different then what you posted. Namely, by default it is:

```
@paragraph{This is me typing some stuff and then I want two @bold{bold
  words}. I don't want the "word" to appear way right.}
```

Because you are it two nested at-exps when that happens. While I absolutely agree that this is a nice thing to have it format that example like that, It does feel like this to me:

```
(quote (this is a sentence that (is built
  entirely using) quote))
```

On the other hand, I guess you could make the argument that this might be what you want if you were building strings. So fair enough.

**Comment by LeifAndersen** [comment]
Actually, after thinking about it a bit more, I'm okay with the indentation you used. (Obviously with one extra space).

**Comment by rfindler** [comment]
Oh wait: I did it wrong. You're absolutely right that two spaces is what it is currently intended to do.

It does that for different examples, basically because it can't tell if it has a function like @bold or if has something where one generally doesn't treat it as part of a paragraph, but instead writes things like:

```
@function1{
 @function2{
  some stuff
 }
}
```

**Comment by rfindler** [comment]
PS: I think it's great that someone looks more closely at that code. There is a good set of test cases, so I'm optimistic that we won't regress :)

**Comment by elibarzilay** [comment]
Just a general comment about the syntax (that is, what the reader reads, not the semantics): it is generally intended to be accomodating for all kinds of tex-like uses, which is why Robby's `@bold` example is indented properly.  But in general, there is obviously difference between those kind of things, and things like `@defproc`, so a full indentation might actully depend on the semantics of the actual form that gets used.

Also, in case it wasn't obvious, since this is generally similar to tex variants, Emacs can be used for comparisons and inspiration too...  (In one of its tex/latex/infotex modes.)

**Comment by rfindler** [comment]
So far, the indentation mode for the at-reader has managed to avoid special casing things in the way that the Racket-mode indentation hasn't (it looks at keywords (without considering their lexical content)). 

Since we have the chance to establish an indentation mode that doesn't require this, I thought it best to try so we don't have to suffer with bad indentation like this:

```
#lang racket
(let ([steps '(begin
                middle-bits
                end)])
  (for ([step (in-list steps)])
    (printf "I'm doing the ~a step!\n" step)))
```

**Comment by elibarzilay** [comment]
Yeah, I know...  And I do think that it's probably not going to be as bad if the indentation rules ignore the command part -- if only because the rules that people might have can vary much more in scribble than in code.  So it's kind of a suggestion thing, to-be-adjusted by the author anyway, rather than what you get in racket code where diverging from the indentation is much more often a bad idea.

**Comment by default-kramer** [comment]
This thread seems to be the best collection of at-exp indentation issues, so I'll add the one I just found here. This is indented with ctrl+i in DrRacket.

```
@(begin
   (define (x)
     (begin
       @defproc*[([(a) b]
                  [(a) b])]{
 blah blah
}
                           ; way over here!
                           (void "something else")
                           )))
```

If you remove the second `begin` then it works as expected:

```
@(begin
   (define (x)
     @defproc*[([(a) b]
                [(a) b])]{
 blah blah
}
     ; all good!
     (void "something else")
     ))
```
